src/main/java/frc/robot/subsystems/drive/TalonSwerveModule.java

```diff
@@ -0,0 +1,171 @@
+    toApply.CurrentLimits.StatorCurrentLimit =
+        TalonDriveVars.TURNING_MOTOR_STATOR_TELEOP_CURRENT_LIMIT_AMPS;
+    toApply.Slot0.kP = TalonDriveVars.TURNING_P;
+    toApply.Slot0.kI = TalonDriveVars.TURNING_I;
+    toApply.Slot0.kD = TalonDriveVars.TURNING_D;
+    toApply.Slot0.kV = TalonDriveVars.TURNING_FF;
+    toApply.Feedback.FeedbackRemoteSensorID = turningTalonEncoder.getDeviceID();
+    toApply.Feedback.FeedbackSensorSource = FeedbackSensorSourceValue.RemoteCANcoder;
+    turningTalon.getConfigurator().apply(toApply);
+  }
+
+  /** Ensures the value a is in [0, b) */
+  public static double mod(double a, double b) {
+    double r = a % b;
+    if (r < 0) {
+      r += b;
+    }
+    return r;
+  }
+
+  /** Consider zeroing based on absolute values */
+  public void considerZeroingEncoder(){
+    if(Math.abs(turningTalonEncoder.getAbsolutePosition().getValueAsDouble()) < 0.01){
+      return;
+    }
+    if(Math.abs(turningTalonEncoder.getAbsolutePosition().getValueAsDouble()-mod(turningTalon.getPosition().getValueAsDouble(), 1)) < TalonDriveVars.TURNING_ENCODER_ZERO_THRESHOLD_ROTATION){
+      turningTalon.setPosition(turningTalonEncoder.getAbsolutePosition().getValueAsDouble());
+    }
+  }
+
+  /** Get the current state of the module */
+  public SwerveModuleState getState(){
+    return new SwerveModuleState(
+      drivingTalon.getVelocity().getValueAsDouble(),
+      new Rotation2d(turningTalonEncoder.getAbsolutePosition().getValueAsDouble())
+    );
+  }
+
+  /** Get the current position of the module */
+  public SwerveModuleState getPosition(){
+    return new SwerveModuleState(
+      drivingTalon.getPosition().getValueAsDouble(),
+      new Rotation2d(turningTalonEncoder.getAbsolutePosition().getValueAsDouble())
+    );
+  }
+
+  /** Applies slew rate. */
+  public double getDesiredVelocityMps(double inputVelocityMps) {
+    // Allow any decrease in desired speed
+    final double prevDesiredVelocityMps = desiredState.speedMetersPerSecond;
+    if (Math.abs(inputVelocityMps) < Math.abs(prevDesiredVelocityMps)) {
+      return inputVelocityMps;
+    }
+
+    // If the change is less than the max accel, allow it
+    final double maxAccelMpss = 15.0;
+    final double loopTimeS = 0.02;
+    final double maxVelChangeMps = maxAccelMpss * loopTimeS;
+    final double velChangeMps = inputVelocityMps - prevDesiredVelocityMps;
+    if (Math.abs(velChangeMps) < maxVelChangeMps) {
+      return inputVelocityMps;
+    }
+
+    // Clamp to max allowed change
+    final double allowedChangeMps = MathUtil.clamp(velChangeMps, -maxVelChangeMps, maxVelChangeMps);
+    return prevDesiredVelocityMps + allowedChangeMps;
+  }
+
+  /** Set the desired state for the module */
+  public void setDesiredState(SwerveModuleState inputState, boolean overrideSlew){
+    inputState =
+        SwerveModuleState.optimize(
+            inputState, new Rotation2d(turningTalonEncoder.getAbsolutePosition().getValueAsDouble()));
```

Comment:
again, shouldnt this be based off the relative encoder?

Reply:
done